src/Components/Components/src/Lifetime/ComponentApplicationState.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// The state for the component and services of a components application.
+    /// </summary>
+    public class ComponentApplicationState
+    {
+        private IDictionary<string, byte[]>? _existingState;
+        private readonly IDictionary<string, byte[]> _currentState;
+        private readonly List<Func<Task>> _registeredCallbacks;
+
+        internal ComponentApplicationState(
+            IDictionary<string, byte[]> currentState,
+            List<Func<Task>> pauseCallbacks)
+        {
+            _currentState = currentState;
+            _registeredCallbacks = pauseCallbacks;
+        }
+
+        internal void InitializeExistingState(IDictionary<string, byte[]> existingState)
+        {
+            if (_existingState != null)
+            {
+                throw new InvalidOperationException("ComponentApplicationState already initialized.");
+            }
+            _existingState = existingState ?? throw new ArgumentNullException(nameof(existingState));
+        }
+
+        /// <summary>
+        /// Registers a callback that will be invoked when the component application is being paused.
+        /// </summary>
+        /// <param name="callback">The <see cref="Func{TResult}"/> to invoke.</param>
+        public void RegisterOnPersistingCallback(Func<Task> callback)
```

Comment:
What are the use cases for this? I don't see it used in the basic example, unless I missed something.

Reply:
They are in the sample I had when originally made the changes, I took the new sample project of the PR so as to not distract the review. The functionality is for services and other more complex components.

```
public class MyService
{
    public MyService(ComponentApplicationState cas)
    {
        cas.RegisterOnPersistingCallback(SaveServiceState);
    }

    private Task SaveServiceState()
    {
        ...
    }
}
```

I'll add samples in when I update the PR to address feedback and add proper tests